sdks/python/apache_beam/typehints/schemas.py

```diff
@@ -647,9 +648,36 @@ def _from_typing(cls, typ):
                                     ('micros', np.int64)])
 
 
+@LogicalType.register_logical_type
+class DateTimeLogicalType(NoArgumentLogicalType[Timestamp, np.int64]):
```

Comment:
Do we need this? I thought the solution here was to use the MicrosInstant logical type

Reply:
ah messed up local virtualenv. Remove `@LogicalType.register_logical_type` or the entire class here causes the following error:
```
Traceback (most recent call last):
  File "...virtualenv/py38beam/lib/python3.8/site-packages/apache_beam/typehints/schemas.py", line 408, in typing_from_runner_api
    field_py_type = self.typing_from_runner_api(field.type)
  File "...virtualenv/py38beam/lib/python3.8/site-packages/apache_beam/typehints/schemas.py", line 362, in typing_from_runner_api
    base = self.typing_from_runner_api(base_type)
  File "...virtualenv/py38beam/lib/python3.8/site-packages/apache_beam/typehints/schemas.py", line 447, in typing_from_runner_api
    return LogicalType.from_runner_api(
  File "...virtualenv/py38beam/lib/python3.8/site-packages/apache_beam/typehints/schemas.py", line 621, in from_runner_api
    raise ValueError(
ValueError: No logical type registered for URN 'beam:logical_type:datetime:v1'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
...
  File "...virtualenv/py38beam/lib/python3.8/site-packages/apache_beam/typehints/schemas.py", line 179, in typing_from_runner_api
    return SchemaTranslation(
  File "...virtualenv/py38beam/lib/python3.8/site-packages/apache_beam/typehints/schemas.py", line 412, in typing_from_runner_api
    raise ValueError(
ValueError: Failed to decode schema due to an issue with Field proto:

name: "f_timestamp"
type {
  nullable: true
  logical_type {
    urn: "beam:logical_type:datetime:v1"
    representation {
      atomic_type: INT64
    }
  }
}
id: 1
encoding_position: 1

```
Still need this to register the URN.

Reply:
I'm confused though, how did we get a field with URN `beam:logical_type:datetime:v1`? Based on the change in JdbcUtil I thought the intention here was to map JDBC timestamp columns to the MicrosInstant logical type, with URN `beam:logical_type:micros_instant:v1`, in order to avoid the proiblematic InstantCoder in DATETIME.

Reply:
The change in JdbcUtil and Python are for different purposes. Basically there are two things missing for java<->portable timestamp: (1) Java JDBCIO support of portable timestamp; (2) Python support for java timestamp.

The JdbcUtil change is for write to jdbc. It adds support in Java's JDBCIO of converting portable timestamp type (beam:logical_type:micros_instant:v1) to sql timestamp;

The Python coder and schema change is for read from jdbc. It adds supports in Python's RowCoder of identifying (by schema change) and decoding (by coder change) the java timestamp type (datetime) to language timestamp. The field 'beam:logical_type:datetime:v1' is passed from java [here](https://github.com/apache/beam/blob/4799828f75309fd494e14eab5af40559e9a140a3/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/SchemaTranslation.java#L201) when the row contains datetime field type.

Reply:
I tried to change coder for Instant in java side but it did not quite work (see comment herehttps://github.com/apache/beam/issues/19817#issuecomment-1204222164) and ended up adding support for InstantCoder coded timestamp in Python side. I understand this is less elegant but it does not touch java Instant coders which is used everywhere not only in schema.

Reply:
Can we make portable timestamp and java timestamp the same type? I think it would be preferable to produce micros_instant for the timestamp type in the JDBC read. I'd really like to avoid having the Java's DATETIME primitive leak into Python.

Reply:
I agree this is preferable in terms of consistency. To do this we will need to make distinction of timestamp in schema based IOs (including JDBC) and windowing timestamp, making the former produce micros_instant instead of joda.Instant, which is similar to the approach here #11456. However given the scope of breaking changes this may leave as next major version change...

If I understood correctly it's not leaking Java's DATETIME primitive (Schema.FieldType DATETIME), the schema passed to python is the logical type `beam:logical_type:datetime:v1` that not yet implemented in python. However the coder that used to decode this type is actually available in python which is TimestampCoder. This is still in line with [Portable Beam Schema design doc](https://docs.google.com/document/d/1uu9pJktzT_O3DxGd1-Q2op4nRk4HekIZbzi-0oTAips/edit#heading=h.w9oanj2p9iih).

Reply:
Could there be a new logical type then for use in JDBC read and write that uses millisecond precision, and maps to joda.Instant?

Reply:
Now renamed this usage to millis_instant logical type